src/main.cpp

```diff
@@ -1014,22 +1086,31 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
         }
     }
 
+    /* Special handling for xversion messages, as they are optional but still
+     have to be properly sequenced to be at the beginning of a connection. So
+     if anything but an xversion comes in whilst in
+     SENT_VERACK_READY_FOR_POTENTIAL_XVERSION, transition to READY state and
+     thus disallow any further incoming xversions. */
+    if (pfrom->state_incoming == ConnectionStateIncoming::SENT_VERACK_READY_FOR_POTENTIAL_XVERSION &&
```

Comment:
If you look at VERACK, XVERSION, BUVERSION, and FILTERSIZEXTHIN are all sent.  Due to the multi-threaded processing, these could be processed in any order. Therefore, you need make this if statement tolerant of BUVERSION and FILTERSIZEXTHIN messages.  I think that doing so still constrains XVERSION to the beginning of the protocol, so its ok.